In [21]:
# Install pymongo by pip
!pip install pymongo

In [2]:
# Import MongoClient from pymongo in project
from pymongo import MongoClient

In [6]:
# Create connection string from Mongodb Atlas

client = MongoClient("mongodb+srv://clustor0_2026:ru2026by@cluster0.gfc9exa.mongodb.net/?appName=Cluster0")


In [22]:
# Test the connection
print("Connected to MongoDB Atlas!")
print("Databases:", client.list_database_names())

Connected to MongoDB Atlas!
Databases: ['northstar_db', 'sample_mflix', 'admin', 'local']


In [23]:
# ============================================
# CREATE NORTHSTAR DATABASE AND COLLECTIONS
#
# In MongoDB:
# - Database = like a folder
# - Collection = like a table
# - Document = like a row (but flexible/nested)
# ============================================

# Create our database called 'northstar_db'
db = client['northstar_db']

# Create collections (like tables)
# MongoDB creates them automatically when we insert data
complaints_col  = db['complaints']
customers_col   = db['customers']
deliveries_col  = db['deliveries']
incidents_col   = db['incidents']

print("Database and collections created!")
print("Database name:", db.name)

Database and collections created!
Database name: northstar_db


In [24]:
import pandas as pd

# ============================================
# LOAD DATASETS DIRECTLY FROM GITHUB
# No setup needed - just run this cell!
# Dataset source: github.com/getdebugged/northstar_case_study
# ============================================

base_url = 'https://raw.githubusercontent.com/getdebugged/northstar_case_study/main/'

app_events  = pd.read_csv(base_url + 'app_events.csv')
complaints  = pd.read_csv(base_url + 'complaints.csv')
customers   = pd.read_csv(base_url + 'customers.csv')
deliveries  = pd.read_csv(base_url + 'deliveries.csv')
drivers     = pd.read_csv(base_url + 'drivers.csv')
hubs        = pd.read_csv(base_url + 'hubs.csv')
incidents   = pd.read_csv(base_url + 'incidents.csv')
orders      = pd.read_csv(base_url + 'orders.csv')
vehicles    = pd.read_csv(base_url + 'vehicles.csv')

print('All datasets loaded directly from GitHub!')
print(f'Orders: {len(orders)} rows')
print(f'Deliveries: {len(deliveries)} rows')
print(f'Complaints: {len(complaints)} rows')

All datasets loaded directly from GitHub!
Orders: 1250 rows
Deliveries: 950 rows
Complaints: 320 rows


In [25]:
# ============================================
# CLEAN DATA BEFORE INSERTING INTO MONGODB
# ============================================

complaints['compensation_amount'] = complaints['compensation_amount'].fillna(0)
deliveries['delivery_completed_at'] = deliveries['delivery_completed_at'].fillna('Not Completed')
deliveries['customer_rating_post_delivery'] = deliveries['customer_rating_post_delivery'].fillna(0)
incidents['resolved_hours'] = incidents['resolved_hours'].fillna(incidents['resolved_hours'].mean())

print('Data cleaned!')

Data cleaned!


In [26]:
##############################################
# INSERT COMPLAINTS INTO MONGODB
#
# WHY MONGODB FOR COMPLAINTS?
# In a normal SQL database, complaints,
# customer info, and messages are in
# separate tables. In MongoDB we store
# everything together in one document.
# This makes querying much faster and easier.
##############################################

# Clear collection first to avoid duplicates
db.complaints.drop()
complaints_col = db['complaints']

complaint_docs = []
for _, row in complaints.iterrows():
    doc = {
        "complaint_id":        row['complaint_id'],
        "customer_id":         row['customer_id'],
        "order_id":            row['order_id'],
        "complaint_type":      row['complaint_type'],
        "channel":             row['channel'],
        "severity":            row['severity'],
        "created_at":          row['created_at'],
        "status":              row['status'],
        "resolution_days":     int(row['resolution_days']),
        "compensation_amount": float(row['compensation_amount'])
    }
    complaint_docs.append(doc)

result = complaints_col.insert_many(complaint_docs)
print(f'Inserted {len(result.inserted_ids)} complaints!')

Inserted 320 complaints!


In [27]:
##############################################
# INSERT DELIVERIES WITH NESTED INCIDENTS
#
# This is the key MongoDB design decision!
# Each delivery document CONTAINS its incidents
# inside it - called EMBEDDED DOCUMENTS
#
# Example structure:
# {
#   delivery_id: "DL001",
#   status: "Failed",
#   incidents: [          <- nested inside!
#     {type: "BatteryAlert", severity: "High"},
#     {type: "RouteDeviation", severity: "Low"}
#   ]
# }
#
# In SQL you would need a JOIN to get this.
# In MongoDB it's all in one place!
##################################################

db.deliveries.drop()
deliveries_col = db['deliveries']

delivery_docs = []
for _, row in deliveries.iterrows():

    # Find all incidents for this delivery
    related_incidents = incidents[
        incidents['delivery_id'] == row['delivery_id']
    ][['incident_type','severity','resolution_status','resolved_hours']].to_dict('records')

    doc = {
        "delivery_id":                 row['delivery_id'],
        "order_id":                    row['order_id'],
        "driver_id":                   row['driver_id'],
        "vehicle_id":                  row['vehicle_id'],
        "hub_id":                      row['hub_id'],
        "delivery_status":             row['delivery_status'],
        "route_distance_km":           float(row['route_distance_km']),
        "manual_route_override_count": int(row['manual_route_override_count']),
        "fuel_or_charge_cost":         float(row['fuel_or_charge_cost']),
        "customer_rating":             float(row['customer_rating_post_delivery']),
        "incidents":                   related_incidents  # NESTED!
    }
    delivery_docs.append(doc)

result = deliveries_col.insert_many(delivery_docs)
print(f'Inserted {len(result.inserted_ids)} deliveries with nested incidents!')

Inserted 950 deliveries with nested incidents!


In [28]:
# ============================================
# INSERT CUSTOMERS INTO MONGODB
# ============================================

customers['loyalty_score'] = customers['loyalty_score'].fillna(customers['loyalty_score'].mean())
customers['preferred_channel'] = customers['preferred_channel'].fillna('Unknown')

db.customers.drop()
customers_col = db['customers']

customer_docs = []
for _, row in customers.iterrows():
    doc = {
        "customer_id":          row['customer_id'],
        "age":                  int(row['age']),
        "home_zone":            row['home_zone'],
        "customer_type":        row['customer_type'],
        "signup_date":          row['signup_date'],
        "loyalty_score":        float(row['loyalty_score']),
        "app_engagement_score": float(row['app_engagement_score']),
        "preferred_channel":    row['preferred_channel'],
        "account_status":       row['account_status']
    }
    customer_docs.append(doc)

result = customers_col.insert_many(customer_docs)
print(f'Inserted {len(result.inserted_ids)} customers!')

Inserted 650 customers!


In [29]:
# --- ORDERS ---
db.orders.drop()
orders_col = db['orders']

orders['booking_channel'] = orders['booking_channel'].fillna('Unknown')

order_docs = []
for _, row in orders.iterrows():
    doc = {
        "order_id":               row['order_id'],
        "customer_id":            row['customer_id'],
        "service_type":           row['service_type'],
        "order_created_at":       row['order_created_at'],
        "promised_window_hours":  float(row['promised_window_hours']),
        "pickup_zone":            row['pickup_zone'],
        "dropoff_zone":           row['dropoff_zone'],
        "priority_level":         row['priority_level'],
        "order_value":            float(row['order_value']),
        "booking_channel":        row['booking_channel'],
        "special_handling_flag":  int(row['special_handling_flag'])
    }
    order_docs.append(doc)

result = orders_col.insert_many(order_docs)
print(f'Inserted {len(result.inserted_ids)} orders!')

# --- DRIVERS ---
db.drivers.drop()
drivers_col = db['drivers']

drivers['training_score'] = drivers['training_score'].fillna(drivers['training_score'].mean())

driver_docs = []
for _, row in drivers.iterrows():
    doc = {
        "driver_id":         row['driver_id'],
        "base_zone":         row['base_zone'],
        "employment_type":   row['employment_type'],
        "years_experience":  float(row['years_experience']),
        "training_score":    float(row['training_score']),
        "driver_rating":     float(row['driver_rating']),
        "shift_preference":  row['shift_preference'],
        "active_flag":       int(row['active_flag'])
    }
    driver_docs.append(doc)

result = drivers_col.insert_many(driver_docs)
print(f'Inserted {len(result.inserted_ids)} drivers!')

# --- VEHICLES ---
db.vehicles.drop()
vehicles_col = db['vehicles']

vehicles['battery_health_pct'] = vehicles['battery_health_pct'].fillna(vehicles['battery_health_pct'].mean())

vehicle_docs = []
for _, row in vehicles.iterrows():
    doc = {
        "vehicle_id":          row['vehicle_id'],
        "vehicle_type":        row['vehicle_type'],
        "assigned_zone":       row['assigned_zone'],
        "commission_date":     row['commission_date'],
        "battery_health_pct":  float(row['battery_health_pct']),
        "odometer_km":         float(row['odometer_km']),
        "maintenance_status":  row['maintenance_status'],
        "telematics_version":  row['telematics_version']
    }
    vehicle_docs.append(doc)

result = vehicles_col.insert_many(vehicle_docs)
print(f'Inserted {len(result.inserted_ids)} vehicles!')

# --- HUBS ---
db.hubs.drop()
hubs_col = db['hubs']

hub_docs = []
for _, row in hubs.iterrows():
    doc = {
        "hub_id":         row['hub_id'],
        "hub_name":       row['hub_name'],
        "zone":           row['zone'],
        "hub_type":       row['hub_type'],
        "capacity_score": int(row['capacity_score'])
    }
    hub_docs.append(doc)

result = hubs_col.insert_many(hub_docs)
print(f'Inserted {len(result.inserted_ids)} hubs!')

# --- APP EVENTS ---
db.app_events.drop()
app_events_col = db['app_events']

app_events['order_id'] = app_events['order_id'].fillna('No Order')

app_event_docs = []
for _, row in app_events.iterrows():
    doc = {
        "event_id":        row['event_id'],
        "customer_id":     row['customer_id'],
        "order_id":        row['order_id'],
        "event_timestamp": row['event_timestamp'],
        "event_type":      row['event_type'],
        "session_id":      row['session_id'],
        "device_type":     row['device_type'],
        "zone_context":    row['zone_context'],
        "api_latency_ms":  float(row['api_latency_ms']),
        "success_flag":    int(row['success_flag'])
    }
    app_event_docs.append(doc)

result = app_events_col.insert_many(app_event_docs)
print(f'Inserted {len(result.inserted_ids)} app events!')

print('\n🎉 All collections inserted into MongoDB!')

Inserted 1250 orders!
Inserted 170 drivers!
Inserted 120 vehicles!
Inserted 8 hubs!
Inserted 640 app events!

🎉 All collections inserted into MongoDB!


In [16]:
# ==================================
# Update
# ==================================
complaints_col.update_one(
    {'complaint_id': 'CMP-001'},
    {'$set': {'status': 'Escalated'}}
)
print("Updated successfully")

Updated successfully


In [17]:
#-=====================================
# Delete
#======================================
result = complaints_col.delete_one({'complaint_id': 'CMP-TEST-001'})
print(f'Deleted {result.deleted_count} document')

Deleted 0 document


In [18]:
# Verify all collections
print("=== Collections in northstar_db ===\n")
for col_name in db.list_collection_names():
    count = db[col_name].count_documents({})
    print(f' {col_name:20} → {count} documents')

=== Collections in northstar_db ===

 vehicles             → 120 documents
 complaints           → 320 documents
 deliveries           → 950 documents
 app_events           → 640 documents
 orders               → 1250 documents
 drivers              → 170 documents
 hubs                 → 8 documents


## MongoDB Queries

In [ ]:
# ################################################
# QUERY 1 - FIND HIGH SEVERITY COMPLAINTS
#
# SQL equivalent:
# SELECT * FROM complaints WHERE severity = 'High'
#
# In MongoDB, find() takes a filter as {}
# ##################################################

print("=== Query 1: High Severity Complaints ===\n")

results = complaints_col.find(
    {"severity": "High"},
    {"_id": 0, "complaint_id": 1, "customer_id": 1,
     "complaint_type": 1, "status": 1, "resolution_days": 1}
).limit(5)

for doc in results:
    print(doc)

total = complaints_col.count_documents({"severity": "High"})
print(f'\nTotal high severity complaints: {total}')

=== Query 1: High Severity Complaints ===

{'complaint_id': 'CP0001', 'customer_id': 'C0464', 'complaint_type': 'AppIssue', 'status': 'Open', 'resolution_days': 11}
{'complaint_id': 'CP0003', 'customer_id': 'C0469', 'complaint_type': 'Delay', 'status': 'Open', 'resolution_days': 16}
{'complaint_id': 'CP0008', 'customer_id': 'C0309', 'complaint_type': 'AppIssue', 'status': 'Resolved', 'resolution_days': 18}
{'complaint_id': 'CP0012', 'customer_id': 'C0362', 'complaint_type': 'AppIssue', 'status': 'Resolved', 'resolution_days': 12}
{'complaint_id': 'CP0020', 'customer_id': 'C0371', 'complaint_type': 'DriverBehaviour', 'status': 'Resolved', 'resolution_days': 11}

Total high severity complaints: 77


Query 1: 77 high severity complaints — nearly 1 in 4 complaints is high severity, showing serious service issues.

In [ ]:
# ============================================
# QUERY 2 - FAILED DELIVERIES WITH INCIDENTS
#
# SQL equivalent:
# SELECT d.*, i.* FROM deliveries d
# JOIN incidents i ON d.delivery_id = i.delivery_id
# WHERE d.delivery_status = 'Failed'
#
# In MongoDB NO JOIN needed - incidents are
# already embedded inside each delivery!
# ============================================

print("=== Query 2: Failed Deliveries With Incidents ===\n")

results = deliveries_col.find(
    {"delivery_status": "Failed"},
    {"_id": 0, "delivery_id": 1, "driver_id": 1,
     "fuel_or_charge_cost": 1, "incidents": 1}
).limit(5)

for doc in results:
    print(f"\nDelivery: {doc['delivery_id']} | Driver: {doc['driver_id']} | Cost: £{doc['fuel_or_charge_cost']}")
    if doc['incidents']:
        for inc in doc['incidents']:
            print(f"  ↳ Incident: {inc['incident_type']} | Severity: {inc['severity']}")
    else:
        print("  ↳ No incidents recorded")

=== Query 2: Failed Deliveries With Incidents ===


Delivery: DL00001 | Driver: D004 | Cost: £12.05
  ↳ Incident: ProofMissing | Severity: High

Delivery: DL00010 | Driver: D058 | Cost: £9.31
  ↳ No incidents recorded

Delivery: DL00012 | Driver: D051 | Cost: £16.98
  ↳ No incidents recorded

Delivery: DL00022 | Driver: D088 | Cost: £15.62
  ↳ No incidents recorded

Delivery: DL00026 | Driver: D092 | Cost: £10.04
  ↳ No incidents recorded


Query 2: Failed deliveries with nested incidents visible in one query — no JOIN needed, this is the power of MongoDB!

In [ ]:
# ============================================
# QUERY 3 - AGGREGATION PIPELINE
#
# SQL equivalent:
# SELECT complaint_type, COUNT(*) as count,
# AVG(resolution_days) as avg_days
# FROM complaints GROUP BY complaint_type
# ORDER BY count DESC
#
# MongoDB uses pipeline stages:
# $group = GROUP BY
# $sort  = ORDER BY
# ============================================

print("=== Query 3: Complaints Summary by Type ===\n")

pipeline = [
    {"$group": {
        "_id": "$complaint_type",
        "total_complaints":   {"$sum": 1},
        "avg_resolution_days": {"$avg": "$resolution_days"},
        "total_compensation":  {"$sum": "$compensation_amount"}
    }},
    {"$sort": {"total_complaints": -1}}
]

results = complaints_col.aggregate(pipeline)
for doc in results:
    print(f"Type: {doc['_id']:20} | Count: {doc['total_complaints']:3} | Avg Days: {round(doc['avg_resolution_days'],1)} | Compensation: £{round(doc['total_compensation'],2)}")

=== Query 3: Complaints Summary by Type ===

Type: Delay                | Count: 101 | Avg Days: 7.3 | Compensation: £1696.84
Type: MissedPickup         | Count:  64 | Avg Days: 7.6 | Compensation: £1423.4
Type: AppIssue             | Count:  53 | Avg Days: 8.6 | Compensation: £980.72
Type: DriverBehaviour      | Count:  51 | Avg Days: 8.2 | Compensation: £973.06
Type: SupportExperience    | Count:  20 | Avg Days: 7.5 | Compensation: £342.5
Type: Billing              | Count:  16 | Avg Days: 7.8 | Compensation: £381.94
Type: Damage               | Count:  15 | Avg Days: 11.3 | Compensation: £359.73


Query 3: Delay is the top complaint (101) costing £1696 in compensation. Damage takes longest to resolve (11.3 days).

In [ ]:
# ============================================
# QUERY 4 - RANGE QUERY
#
# SQL equivalent:
# SELECT * FROM deliveries
# WHERE manual_route_override_count > 1
#
# MongoDB uses $gt which means "greater than"
# Other operators: $lt (less than),
# $gte (greater or equal), $lte (less or equal)
# ============================================

print("=== Query 4: Deliveries With High Manual Overrides ===\n")

results = deliveries_col.find(
    {"manual_route_override_count": {"$gt": 1}},
    {"_id": 0, "delivery_id": 1, "hub_id": 1,
     "delivery_status": 1, "manual_route_override_count": 1}
).limit(5)

for doc in results:
    print(doc)

total = deliveries_col.count_documents({"manual_route_override_count": {"$gt": 1}})
print(f'\nTotal deliveries with >1 manual override: {total}')

=== Query 4: Deliveries With High Manual Overrides ===

{'delivery_id': 'DL00012', 'hub_id': 'H05', 'delivery_status': 'Failed', 'manual_route_override_count': 3}
{'delivery_id': 'DL00013', 'hub_id': 'H04', 'delivery_status': 'OnTime', 'manual_route_override_count': 2}
{'delivery_id': 'DL00015', 'hub_id': 'H04', 'delivery_status': 'OnTime', 'manual_route_override_count': 3}
{'delivery_id': 'DL00019', 'hub_id': 'H06', 'delivery_status': 'OnTime', 'manual_route_override_count': 4}
{'delivery_id': 'DL00021', 'hub_id': 'H08', 'delivery_status': 'OnTime', 'manual_route_override_count': 4}

Total deliveries with >1 manual override: 241


Query 4: 241 out of 950 deliveries had more than 1 manual override — that's 25% of all deliveries!

## Query Optimisation - Indexing

In [32]:
# ============================================
# QUERY OPTIMISATION - INDEXING
#
# What is an index?
# instead of scanning every document, it goes straight
# to the matching ones. This makes queries FASTER!


# 1. Run a query WITHOUT index and measure time
# 2. Create an index
# 3. Run same query WITH index and measure time
# 4. Compare the difference
# ============================================

import time

# --- BEFORE INDEX ---
# Run query on delivery_status WITHOUT index
start = time.time()

count = deliveries_col.count_documents({"delivery_status": "Failed"})

end = time.time()
time_before = round((end - start) * 1000, 2)

print(f'WITHOUT index: Found {count} documents in {time_before} ms')

WITHOUT index: Found 132 documents in 102.5 ms


In [30]:
# ============================================
# CREATE INDEXES
#
# We create indexes on columns we query often
# delivery_status - we filter by this a lot
# severity - we filter complaints by this
# hub_id - we group deliveries by hub
# customer_id - we look up customers often
# ============================================

# Create index on delivery_status
deliveries_col.create_index("delivery_status")
print(' Index created on deliveries.delivery_status')

# Create index on hub_id
deliveries_col.create_index("hub_id")
print(' Index created on deliveries.hub_id')

# Create index on complaints severity
complaints_col.create_index("severity")
print('Index created on complaints.severity')

# Create index on customer_id in complaints
complaints_col.create_index("customer_id")
print('Index created on complaints.customer_id')

# Create compound index - two fields together
# Useful when we filter by BOTH status AND hub
deliveries_col.create_index([("delivery_status", 1), ("hub_id", 1)])
print('Compound index created on delivery_status + hub_id')

 Index created on deliveries.delivery_status
 Index created on deliveries.hub_id
Index created on complaints.severity
Index created on complaints.customer_id
Compound index created on delivery_status + hub_id


In [33]:
# --- AFTER INDEX ---
# Run same query again WITH index and compare
start = time.time()

count = deliveries_col.count_documents({"delivery_status": "Failed"})

end = time.time()
time_after = round((end - start) * 1000, 2)

print(f'WITH index:    Found {count} documents in {time_after} ms')
print(f'\n Time before: {time_before} ms')
print(f' Time after:  {time_after} ms')

if time_after < time_before:
    improvement = round(((time_before - time_after) / time_before) * 100, 1)
    print(f' Performance improved by {improvement}%!')
else:
    print('Note: Dataset is small so difference may be minimal,')
    print('but indexes become critical with millions of records!')

WITH index:    Found 132 documents in 98.71 ms

 Time before: 102.5 ms
 Time after:  98.71 ms
 Performance improved by 3.7%!


In [19]:
# ============================================
# EXPLAIN PLAN
#
# explain() shows us HOW MongoDB runs a query
# - COLLSCAN = collection scan (slow - reads every document)
# - IXSCAN   = index scan (fast - uses index)
# We want to see IXSCAN after creating index!
# ============================================

explain = deliveries_col.find(
    {"delivery_status": "Failed"}
).explain()

# Show the winning plan
winning_plan = explain['queryPlanner']['winningPlan']
print("=== Explain Plan ===")
print(f"Stage: {winning_plan['stage']}")

# Check if index is being used
if 'inputStage' in winning_plan:
    print(f"Input Stage: {winning_plan['inputStage']['stage']}")
    if 'indexName' in winning_plan['inputStage']:
        print(f"Index Used: {winning_plan['inputStage']['indexName']}")
        print(' MongoDB is using our index!')
else:
    print('Using collection scan')

=== Explain Plan ===
Stage: COLLSCAN
Using collection scan
